### ISSUE TITLE
[v2.8] [BUG] Fleet controller can't provision local cluster immediately

### ISSUE THREAD

**Issue by bk201** [rancher/rancher#41848]
## SURE-6506

**Rancher Server Setup**
- Rancher version: v2.7.5-rc5
- Installation option (Docker install/Helm Chart):
   - If Helm Chart, Kubernetes Cluster and version (RKE1, RKE2, k3s, EKS, etc):
- Proxy/Cert Details:

**Information about the Cluster**
- Kubernetes version: v1.25.9+rke2r1
- Cluster Type (Local/Downstream): local
   - If downstream, what type of cluster? (Custom/Imported or specify provider for Hosted/Infrastructure Provider):

**User Information**
- What is the role of the user logged in? (Admin/Cluster Owner/Cluster Member/Project Owner/Project Member/Custom)
  - If custom, define the set of permissions:

**Describe the bug**

We use [rancherd](https://github.com/rancher/rancherd) to bootstrap the Harvester cluster with Rancher. After integrating v2.7.5-rc5, we found the local cluster provisioning doesn't work due to the fleet controller requiring `apiServerURL` to be set (thanks, @Oats87 ). [We added `apiServerURL` manually](https://github.com/rancher/rancherd/pull/31) to fix it.

We observed that the local cluster provision doesn't happen immediately after we set the secret; sometimes, it takes a very long time (up to 10 minutes). Restarting fleet-controller pod makes the cluster provision happen immediately.

**To Reproduce**

Provision a local cluster with rancherd; its basic flow is:
- Deploy an RKE2 server node.
- Deploy Rancher, rancher will deploys rancher-webhook and fleet controller.
- Update secret `local-kubeconfig`. 
    - `apiServerURL`: value of Rancher setting "internal-server-url".
    - `apiServerCA`: value of Rancher setting "internal-cacerts".
- Apply cluster manifests:
  ```
  ---
  apiVersion: v1
  kind: Node
  metadata:
    labels:
      node-role.kubernetes.io/etcd: "true"
    name: node1
  
  ---
  apiVersion: v1
  kind: Namespace
  metadata:
    name: fleet-local
  
  ---
  apiVersion: provisioning.cattle.io/v1
  kind: Cluster
  metadata:
    name: local
    namespace: fleet-local
  spec:
    kubernetesVersion: v1.25.9+rke2r1
    rkeConfig:
      controlPlaneConfig:
        disable:
        - rke2-snapshot-controller
        - rke2-snapshot-controller-crd
        - rke2-snapshot-validation-webhook
  
  ---
  apiVersion: v1
  data:
    agentToken: YWEgYmIgY2M=
    serverToken: YWEgYmIgY2M=
  kind: Secret
  metadata:
    name: local-rke-state
    namespace: fleet-local
  type: rke.cattle.io/cluster-state
  
  ---
  apiVersion: management.cattle.io/v3
  kind: ClusterRegistrationToken
  metadata:
    name: default-token
    namespace: local
  spec:
    clusterName: local
  status:
    token: aa bb cc
  
  ---
  apiVersion: catalog.cattle.io/v1
  kind: ClusterRepo
  metadata:
    name: rancher-stable
  spec:
    url: https://releases.rancher.com/server-charts/stable
  ```
- After a few minutes, we'll see fleet-agent spawned and provisioning the local cluster. This process happens immediately in previous versions (Rancher v2.6.x)

**Result**

Fleet-agent is not spawned immediately to do the rest of the work.

**Expected Result**

Fleet-agent is spawned and provision the local cluster once we create the manifests.

**Screenshots**

**Additional context**

If we restart the fleet-controller pod, the fleet-agent will be deployed immediately.

Logs:
- [rancher.txt](https://github.com/rancher/rancher/files/11752450/rancher.txt)
- [fleet-controller.txt](https://github.com/rancher/rancher/files/11752452/fleet-controller.txt)

We added a [workaround](https://github.com/rancher/rancherd/pull/32) to restart the fleet-controller pod to speed up the process. But it's better to see if Rancher can set the correct API URLs when creating the `local-kubeconfig` secret (before deploying fleet-controller), otherwise, the workaround might break again.

**Comment by bk201** [comment]
I wrote a guide to reproduce the issue, hope it helps: https://github.com/bk201/vagrant-rancherd/blob/repro-rancher-41848/docs/repro-rancher-41848.md

**Comment by ibrokethecloud** [comment]
This seems to impact us during the upgrades as well when rancherd bootstrap is not performed.

fleet-controller stays stuck on waiting for `apiServerURL` to be set. 

In general may be a special case needs to be added when handling import of `local` cluster, where the apiServerURL and apiServerCA is derived from the `kubeconfig` for the local cluster. This will ensure local fleet-agent is only using the rancher svc for registration and not using the `server-url` setting.

I'd be happy to provide the fix, if we are happy with this suggestion?

**Comment by weyfonk** [comment]
Thanks @ibrokethecloud , we have submitted a [PR](https://github.com/rancher/rancher/pull/42649) to Rancher to populate the local cluster's API server URL into its `kubeconfig` secret. Feedback welcome!

**Comment by thehejik** [comment]
### Test report
Successful verified on rancher `v2.8-0b0d677fa3c980137dcc1cd120b203d154728d21-head`.

I was able reproduce the issue on `2.8.3-rc2`:
![image](https://github.com/rancher/rancher/assets/12828077/11049222-0932-4884-a4fe-e06a49b8a7b7) 

On `v2.8-head` after deploying rancher both values for `apiServerCA` and `apiServerURL` keys are present instantly in `local-kubeconfig` secret on local cluster (on `2.8.3-rc2` there is just `token` and `value` present).

It still takes some time to provision local cluster with fleet after initial login to rancher `v2.8-head`, around 90 seconds, but it's way better than in `2.8.3-rc2`, also the error message about missing `apiServerURL` is not present there.

**Comment by thehejik** [comment]
It is verified fixed in `2.8.3-rc3`.

**Comment by Jono-SUSE-Rancher** [comment]
Closing as it was verified by QA.
